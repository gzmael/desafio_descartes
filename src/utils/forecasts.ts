import { Forecast } from 'models/Forecast'

export const getDayByDate = (date: string) =>
  new Intl.DateTimeFormat('pt-br', {
    day: '2-digit',
    timeZone: 'America/Fortaleza',
  }).format(new Date(date))

export const getHourByTimestamp = (time: number) =>
  new Intl.DateTimeFormat('pt-br', {
    hour: '2-digit',
    minute: '2-digit',
    timeZone: 'America/Fortaleza',
  }).format(new Date((time + 10800) * 1000))

export const getDayWeekByDate = (
  date: string,
  style?: 'short' | 'long' | 'narrow',
) =>
  new Intl.DateTimeFormat('pt-br', {
    weekday: style ?? 'long',
    timeZone: 'America/Fortaleza',
  }).format(new Date(date))

export const getWeekAndDaysFromForecasts = (forecasts: Forecast[]) => {
  const dates = [
    ...new Set(forecasts.map((forecast) => forecast.date.split(' ')[0])),
  ]
  return dates.map((date) => {
    return {
      date,
      day: getDayByDate(date + ' 12:00:00'),
      week: getDayWeekByDate(date + ' 12:00:00', 'short'),
    }
  })
}
