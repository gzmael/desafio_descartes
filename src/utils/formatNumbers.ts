export const formatTemperature = (temp: number, maxFrac = 1) => {
  return new Intl.NumberFormat('pt-br', {
    maximumFractionDigits: maxFrac,
  }).format(temp)
}

export const formatMsToKmH = (value: number, digits = 2) => {
  return new Intl.NumberFormat('pt-br', {
    maximumFractionDigits: digits,
  }).format(value * 3.6)
}
