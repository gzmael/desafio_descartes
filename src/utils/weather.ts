import { Weather, WeatherResponse } from 'models/Weathear'
import { formatMsToKmH, formatTemperature } from './formatNumbers'
import { Forecast, ForecastReponse } from 'models/Forecast'

export const getDetailListFromWeather = (weather: Weather) => {
  return [
    {
      title: 'Vento',
      value: formatMsToKmH(weather.wind) + ' km/h',
    },
    {
      title: 'Umidade',
      value: weather.humidity + ' %',
    },
    {
      title: 'Sensação',
      value: formatTemperature(weather.feels) + ' ºC',
    },
    {
      title: 'Min - Máx',
      value: `${formatTemperature(weather.min, 0)}ºC - ${formatTemperature(
        weather.max,
        0,
      )}ºC`,
    },
    {
      title: 'Visibilidade',
      value: weather.visibility / 1000 + ' km',
    },
    {
      title: 'Nuvens',
      value: weather.clouds + ' %',
    },
  ]
}

export const getWeatherFromData = (data: WeatherResponse) => {
  const { clouds, main, visibility, weather, wind, sys } = data
  return {
    clouds: clouds.all,
    description: weather[0].description,
    feels: main.feels_like,
    humidity: main.humidity,
    icon: weather[0].icon,
    max: main.temp_max,
    min: main.temp_min,
    temperature: main.temp,
    wind: wind.speed,
    pressure: main.pressure,
    sunrise: sys.sunrise,
    sunset: sys.sunset,
    visibility,
  }
}

export const getForecastsFromData = (data: ForecastReponse): Forecast[] => {
  return data.list.map((forecast) => {
    const {
      clouds,
      dt_txt: date,
      main,
      visibility,
      weather,
      wind,
      dt,
    } = forecast
    return {
      clouds: clouds.all,
      description: weather[0].description,
      feels: main.feels_like,
      humidity: main.humidity,
      icon: weather[0].icon,
      max: main.temp_max,
      min: main.temp_min,
      temperature: main.temp,
      wind: wind.speed,
      pressure: main.pressure,
      visibility,
      date,
      timestamp: dt,
    }
  })
}
