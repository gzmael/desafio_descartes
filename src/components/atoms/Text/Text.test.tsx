import { render, screen } from '@testing-library/react'

import { Text, TextProps } from '.'
import { cn } from 'utils/cls-merge'

describe('Component/Atom/Text', () => {
  const defaultClassName = ['leading-6', 'tracking-wider']
  it('should render text with default variant', () => {
    const props: TextProps = {
      children: 'Hello World',
    }

    const classNames = cn([
      defaultClassName,
      'text-base',
      'text-white',
      'font-medium',
    ])

    const { container } = render(<Text {...props} />)

    expect(screen.getByText(/Hello World/i)).toBeInTheDocument()

    expect(container.firstChild).toHaveClass(classNames)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render text with small size variant', () => {
    const props: TextProps = {
      children: 'Small Text',
      size: 'sm',
    }

    const classNames = cn([
      defaultClassName,
      'text-sm',
      'text-white',
      'font-medium',
    ])

    const { container } = render(<Text {...props} />)

    expect(screen.getByText(/Small Text/i)).toBeInTheDocument()

    expect(container.firstChild).toHaveClass(classNames)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render text with medium size and primary color', () => {
    const props: TextProps = {
      children: 'Medium Text',
      type: 'primary',
    }

    const classNames = cn([
      defaultClassName,
      'text-base',
      'text-light-blue-500',
      'font-medium',
    ])

    const { container } = render(<Text {...props} />)

    expect(screen.getByText(/Medium Text/i)).toBeInTheDocument()

    expect(container.firstChild).toHaveClass(classNames)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render text with large size and bold', () => {
    const props: TextProps = {
      children: 'Medium Text',
      type: 'primary',
      size: 'lg',
      weight: 'bold',
    }

    const classNames = cn([
      defaultClassName,
      'text-lg',
      'text-light-blue-500',
      'font-bold',
    ])

    const { container } = render(<Text {...props} />)

    expect(screen.getByText(/Medium Text/i)).toBeInTheDocument()

    expect(container.firstChild).toHaveClass(classNames)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render text with Slot tag', () => {
    const props: TextProps = {
      children: <span>Slot Text</span>,
      asChild: true,
    }

    const classNames = cn([
      defaultClassName,
      'text-base',
      'text-white',
      'font-medium',
    ])

    const { container } = render(<Text {...props} />)

    expect(
      screen.getByText(/Slot Text/, {
        selector: 'span',
      }),
    ).toBeInTheDocument()

    expect(container.firstChild).toHaveClass(classNames)

    expect(container.firstChild).toMatchSnapshot()
  })
})
