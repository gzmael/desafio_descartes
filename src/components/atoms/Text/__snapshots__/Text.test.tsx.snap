// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component/Atom/Text > should render text with Slot tag 1`] = `
<span
  class="tracking-wider text-base text-white font-medium"
>
  Slot Text
</span>
`;

exports[`Component/Atom/Text > should render text with default variant 1`] = `
<p
  class="tracking-wider text-base text-white font-medium"
>
  Hello World
</p>
`;

exports[`Component/Atom/Text > should render text with large size and bold 1`] = `
<p
  class="tracking-wider text-lg text-ligthblue-500 font-bold"
>
  Medium Text
</p>
`;

exports[`Component/Atom/Text > should render text with medium size and primary color 1`] = `
<p
  class="tracking-wider text-base text-ligthblue-500 font-medium"
>
  Medium Text
</p>
`;

exports[`Component/Atom/Text > should render text with small size variant 1`] = `
<p
  class="tracking-wider text-sm text-white font-medium"
>
  Small Text
</p>
`;
